### ISSUE TITLE
Feature request: ability to set our own implementations of global rate limiter

### ISSUE THREAD

**Issue by Alex1304** [Discord4J/Discord4J#531]
**Feature Description:**

Change the current GlobalRateLimiter class into an interface with 3 methods:

```java
public interface GlobalRateLimiter {
	
	public void rateLimitFor(Duration duration);

	public Mono<T> withRateLimiter(Mono<T> stage);

	// Currently known as "long delayNanos()", called in the retry factory.
	// I thought a better signature would be needed since this method would go public.
	public Duration getRemainingRateLimit();
}
```
The default implementation with Semaphores, as well as the upcoming implementation with `reactor-pool` that @quanticc is cooking up, would be implementations of this interface, and would be used by default if no custom implementation is provided. A custom implementation would be supplied via the RouterOptions, like so:

```java
RouterOptions.builder()
	.globalRateLimiter(new MyCustomGlobalRateLimiter())
	.build();
```

This would imply deprecating and removing `.requestParallelism`, since parallelism does only make sense for the default implementation. It could still be provided this way instead:

```java
RouterOptions.builder()
	.globalRateLimiter(new DefaultGlobalRateLimiter(parallelism))
	.build();
```

**Justification:**

It would allow for even more freedom than just setting the request parallelism, as we would be able to have full control on how global rate limits are handled. It would cover even more use cases such as:

- Delaying requests in order to fulfill custom limits (for example if we want our application to make a maximum of X requests per minute, per hour, etc)
- Use cases that needs unbounded request parallelism (currently we can still do `.requestParallelism(Integer.MAX_VALUE)`, but instead of having a Semaphore/ReactorPool of unbounded size, the user could make an implementation that simply doesn't have any resource acquiring logic at all)
- Cases where we want to completely ignore global rate limits, conditionally or permanently (a sort of NoOpGlobalRateLimiter that would make all requests retry immediately even if Discord signals that global rate limit is being reached). Should be used at our own risk of course, lol
- Using a GlobalRateLimiter as a middleware, that triggers an action on each request made (e.g incrementing a counter, saving something in database, etc)

**Comment by quanticc** [comment]
Thanks! This is on the roadmap for v3.1, where we're interested in making core module more friendly across processes.

At the moment to achieve this you could only replace the entire Router implementation which involves some heavy lifting. Allowing custom GRL is the least one should replace to provide inter-process communication.

**Comment by quanticc** [comment]
Implemented this for 3.0.x as it's possible without impacting current bot code.
